daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DaffodilInputContentHandler.scala

```diff
@@ -0,0 +1,180 @@
+  extends DFDL.DaffodilInputContentHandler {
+  private val inputter = new SAXInfosetInputter(this, dp, output)
+  private var unparseResult: DFDL.UnparseResult = _
+  private var infosetEvent: DFDL.SaxInfosetEvent = _
+  private val characterData = new StringBuilder()
+
+  def setUnparseResult(ur: DFDL.UnparseResult): Unit = {
+    unparseResult = ur
+  }
+
+  def getUnparseResult: DFDL.UnparseResult = unparseResult
+
+  override def setDocumentLocator(locator: Locator): Unit = {
+    // do nothing
+  }
+
+  override def startDocument(): Unit = {
+    infosetEvent = new DFDL.SaxInfosetEvent
+    infosetEvent.eventType = One(StartDocument)
+    sendToInputter
+  }
+
+  override def endDocument(): Unit = {
+    if (infosetEvent == null) {
+      infosetEvent = new DFDL.SaxInfosetEvent
+    }
+    infosetEvent.eventType = One(EndDocument)
+    sendToInputter
+  }
+
+  // TODO not sure what to do with the prefix mappings, technically we don't need them since
+  // we have the attributes with startElement to look up any prefixes we need
+  override def startPrefixMapping(prefix: String, uri: String): Unit = {
+    val pre = if (prefix == "") null else prefix
+    infosetEvent.prefixMapping = One(NamespaceBinding(pre, uri, infosetEvent.prefixMapping
+      .orNull))
+  }
+
+  override def endPrefixMapping(prefix: String): Unit = {
+    // do nothing
+  }
+
+  override def startElement(uri: String, localName: String, qName: String, atts: Attributes): Unit = {
+    // we need to check if the characters info is all whitespace, if not, it is an error
+    if (characterData.nonEmpty && characterData.toString().trim.nonEmpty) {
+      throw new IllegalContentWhereEventExpected("Non-whitespace characters in complex " +
+        "Element: " + characterData.toString())
+    } else {
+      // reset since it was whitespace only
+      characterData.setLength(0)
+    }
+
+    if (infosetEvent == null) infosetEvent = new DFDL.SaxInfosetEvent
+    else if (!infosetEvent.isEmpty && infosetEvent.localName.isDefined) {
+      // we started another element while we were in the process of building another
+      // this most likely means the first element was complex, either ways, it is ready
+      // for the inputter queue
+      sendToInputter
+    }
+    // use Attributes to determine xsi:nil value
+    val nilIn = atts.getIndex(XMLConstants.W3C_XML_SCHEMA_INSTANCE_NS_URI, "nil")
+    infosetEvent.nilValue = if (nilIn >= 0) {
+      val nilValue = atts.getValue(nilIn)
+      One(nilValue)
+    } else {
+      Nope
+    }
+    // set localName and namespaceURI
+    if (localName.nonEmpty && uri.nonEmpty) {
+      infosetEvent.namespaceURI = One(uri)
+      infosetEvent.localName = One(localName)
+    } else if (qName.nonEmpty) {
+      val qNameArr = qName.split(":")
+
+      if (qNameArr.length > 1) {
+        infosetEvent.localName = qNameArr.lift(1)
+        val qNamePrefix = qNameArr.lift(0).get
+        val in = atts.getIndex(s"xmlns:$qNamePrefix")
```

Comment:
This implies that the namespace for this prefix is defined on this element. But it could have been defined on some parent element. I think perhaps this is where the prefix mapping comes into place. Perhaps we need to build a NamespaceBuilding, and if we do not get uri + local name, then we have to look up the prefix (which could be empty string) in the NamespaceBinding?